### PR TITLE
allow testing for functional options

### LINKED ISSUE #997: Mock Matcher for functional options
Hi,

In most go libraries nowadays (etcd, kubernets, aws, etc.) the [functional options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) is being used:

```go
c := client.New(client.WithTimeout(), client.WithLabel("bar"))
```

The mocks for these functional options are unfortunately unable to match since the option usually is implemented so:

```go
type opt struct {
  timeout bool
  label string
}

type ClientOption func(*opt)

func WithTimeout() ClientOption {
  return func(o *opt) {
    o.timeout = true
  }
}

func WithLabel(v string) ClientOption {
  return func(o *opt) {
    o.label = v
  }
}

func (c *client) New(opts ...ClientOption) *someClient {
  options := &opt{}
  for _, f := range opts {
    f(options)
  }

  return &someClient{options.timeout, options.label}
}
```

`testify.Mock` matcher will not match `x := WithTimeout()` with `y := WithTimeout()` since these are actually distinct functions. Using `mock.AnythingOfType` defeats the purpose since these options are usually significant and even critical to behavior.

As a workaround currently, I'm using reflection and the `Call.Run()` (alt. `mock.MatchedBy`) function to match these `opts` and log a proper diff (different options, different type, wrong values, etc.). I thought about it and it would've been ideal if we can introduce a matcher type `mock.OptionTypeArgument` or something since the MatchedBy func is inadequate for printing meaningful diffs as it will only true/false info of the failure scenario.

Is this something that the library could/should support? If yes, I can open a PR to elicit the approach.

### PR DESCRIPTION
**Pull request by nbaztec**
## Summary

This PR fixes #997 

In most go libraries nowadays (etcd, kubernets, aws, etc.) the [functional options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) is being used:

```go
c := client.New(client.WithTimeout(), client.WithLabel("bar"))
```

```go
type opt struct {
  timeout bool
  label string
}

type ClientOption func(*opt)

func WithTimeout() ClientOption {
  return func(o *opt) {
    o.timeout = true
  }
}

func WithLabel(v string) ClientOption {
  return func(o *opt) {
    o.label = v
  }
}

func (c *client) New(opts ...ClientOption) *someClient {
  options := &opt{}
  for _, f := range opts {
    f(options)
  }

  return &someClient{options.timeout, options.label}
}
```

## Changes
Adds a new argument type `FunctionalOptions`

## Motivation
This wasn't possible before.

## Related issues
Closes #997 #1006

### CODE REVIEW

**Review by alameddinc — DISMISSED** [review]
I think It should be inserted into the mock library 👍

**Comment by dillonstreator** [comment]
Any update on this? Would love to see functional option testing support.

File: mock/mock.go [hunk 1]
```diff
@@ -798,6 +828,23 @@ func (args Arguments) Diff(objects []interface{}) (string, int) {
 				differences++
 				output = fmt.Sprintf("%s\t%d: FAIL:  type %s != type %s - %s\n", output, i, reflect.TypeOf(t).Name(), reflect.TypeOf(actual).Name(), actualFmt)
 			}
+		} else if reflect.TypeOf(expected) == reflect.TypeOf((*FunctionalOptionsArgument)(nil)) {
+			name := expected.(*FunctionalOptionsArgument).name
+			t := expected.(*FunctionalOptionsArgument).value
+			tName := reflect.TypeOf(t).Name()
+			if name != reflect.TypeOf(actual).String() {
```
Comment (dillonstreator) [inline mock/mock.go:835]: We should be derive the `name` field on `FunctionalOptionsArgument` with a bit more reflection and simplify the `FunctionalOptions()` signature

```suggestion
			t := expected.(*FunctionalOptionsArgument).value

			var name string
			tValue := reflect.ValueOf(t)
			if tValue.Len() > 0 {
				name = "[]" + reflect.TypeOf(tValue.Index(0).Interface()).String()
			}

			tName := reflect.TypeOf(t).Name()
			if name != reflect.TypeOf(actual).String() && tValue.Len() != 0 {
```

File: mock/mock.go [hunk 2]
```diff
@@ -659,6 +665,30 @@ func IsType(t interface{}) *IsTypeArgument {
 	return &IsTypeArgument{t: t}
 }
 
+// FunctionalOptionsArgument is a struct that contains the type and value of an functional option argument
+// for use when type checking.
+type FunctionalOptionsArgument struct {
+	name  string
```
Comment (dillonstreator) [inline mock/mock.go:671]:
```suggestion
```

File: mock/mock.go [hunk 3]
```diff
@@ -659,6 +665,30 @@ func IsType(t interface{}) *IsTypeArgument {
...
+	value interface{}
+}
+
+// String returns the string representation of FunctionalOptionsArgument
+func (f *FunctionalOptionsArgument) String() string {
+	return strings.Replace(fmt.Sprintf("%#v", f.value), "[]interface {}", f.name, 1)
```
Comment (dillonstreator) [inline mock/mock.go:677]:
```suggestion
	var name string
	tValue := reflect.ValueOf(f.value)
	if tValue.Len() > 0 {
		name = "[]" + reflect.TypeOf(tValue.Index(0).Interface()).String()
	}

	return strings.Replace(fmt.Sprintf("%#v", f.value), "[]interface {}", name, 1)
```

File: mock/mock.go [hunk 4]
```diff
@@ -659,6 +665,30 @@ func IsType(t interface{}) *IsTypeArgument {
...
+// FunctionalOptions returns an FunctionalOptionsArgument object containing the functional option type
+// and the values to check of
+//
+// For example:
+// Assert(t, FunctionalOptions("[]foo.FunctionalOption", foo.Opt1(), foo.Opt2()))
+func FunctionalOptions(name string, value ...interface{}) *FunctionalOptionsArgument {
+	return &FunctionalOptionsArgument{
+		name:  name,
+		value: value,
+	}
```
Comment (dillonstreator) [inline mock/mock.go:689]:
```suggestion
func FunctionalOptions(value ...interface{}) *FunctionalOptionsArgument {
	return &FunctionalOptionsArgument{
		value: value,
	}
```

File: mock/mock_test.go [hunk 5]
```diff
@@ -1044,6 +1067,23 @@ func Test_Mock_AssertExpectationsCustomType(t *testing.T) {
 
 }
 
+func Test_Mock_AssertExpectationsFunctionalOptionsType(t *testing.T) {
+
+	var mockedService = new(TestExampleImplementation)
+
+	mockedService.On("TheExampleMethodFunctionalOptions", "test", FunctionalOptions("[]mock.OptionFn", OpNum(1), OpStr("foo"))).Return(nil).Once()
```
Comment (dillonstreator) [inline mock/mock_test.go:1074]:
```suggestion
	mockedService.On("TheExampleMethodFunctionalOptions", "test", FunctionalOptions(OpNum(1), OpStr("foo"))).Return(nil).Once()
```

File: mock/mock_test.go [hunk 6]
```diff
@@ -1044,6 +1067,23 @@ func Test_Mock_AssertExpectationsCustomType(t *testing.T) {
...
+	assert.False(t, mockedService.AssertExpectations(tt))
+
+	// make the call now
+	mockedService.TheExampleMethodFunctionalOptions("test", OpNum(1), OpStr("foo"))
+
+	// now assert expectations
+	assert.True(t, mockedService.AssertExpectations(tt))
+
+}
+
```
Comment (dillonstreator) [inline mock/mock_test.go:1086]: Add test for empty functional options

```suggestion

func Test_Mock_AssertExpectationsFunctionalOptionsType_Empty(t *testing.T) {

	var mockedService = new(TestExampleImplementation)

	mockedService.On("TheExampleMethodFunctionalOptions", "test", FunctionalOptions()).Return(nil).Once()

	tt := new(testing.T)
	assert.False(t, mockedService.AssertExpectations(tt))

	// make the call now
	mockedService.TheExampleMethodFunctionalOptions("test")

	// now assert expectations
	assert.True(t, mockedService.AssertExpectations(tt))

}

```

**Review by dillonstreator — CHANGES_REQUESTED** [review]
@nbaztec thoughts on these changes to simplify the signature?

**Comment by dillonstreator** [comment]
@boyan-soubachov can you please provide feedback on this PR?

**Comment by nbaztec** [comment]
@dillonstreator The suggestions look good, I've incorporated them. 
I was a bit hesitant at first if there could arise a situation during a refactor/upgrade, where the first parameter could be an incorrect type in the assertion, thereby using an incorrect type - but that seems like a non-issue, since the same can happen when specifying a manual string.

**Review by boyan-soubachov — APPROVED** [review]
Thank you :)

File: mock/mock.go [hunk 7]
```diff
@@ -798,6 +832,29 @@ func (args Arguments) Diff(objects []interface{}) (string, int) {
...
+			var name string
+			tValue := reflect.ValueOf(t)
+			if tValue.Len() > 0 {
+				name = "[]" + reflect.TypeOf(tValue.Index(0).Interface()).String()
+			}
+
+			tName := reflect.TypeOf(t).Name()
+			if name != reflect.TypeOf(actual).String() && tValue.Len() != 0 {
+				differences++
+				output = fmt.Sprintf("%s\t%d: FAIL:  type %s != type %s - %s\n", output, i, tName, reflect.TypeOf(actual).Name(), actualFmt)
```
Comment (bhingston-va) [inline mock/mock.go:847]: I believe I'm following the test setup nearly the same (options are from another package) and I noticed 2 things.
1. `tName` is empty for me but `reflect.TypeOf(tValue.Index(0).Interface()).String()` has the same type as `reflect.TypeOf(actual).String()`
2. `name` has the prefix `[]` so I'm hitting this branch

Curious the choice for `tName` and `reflect.TypeOf(actual).Name()` in the log here instead of `name` and `reflect.TypeOf(actual).String()` that is used in the conditional.

It was hard for me to debug as the logs showed empty string is not the same type as my option type. When debugging I found that `name` was really close except for the prefixed `[]`

I might follow up with a PR to fix the logged out line here if it was just an accident.
